runners/java-fn-execution/src/main/java/org/apache/beam/runners/fnexecution/control/DefaultJobBundleFactory.java

```diff
@@ -192,6 +202,28 @@ public static DefaultJobBundleFactory create(
     ImmutableList.Builder<LoadingCache<Environment, WrappedSdkHarnessClient>> caches =
         ImmutableList.builder();
     for (int i = 0; i < count; i++) {
+      final Lock refLock = environmentCacheLocks.get(i);
+      builder = builder.removalListener(
+          notification -> {
+            WrappedSdkHarnessClient client = (WrappedSdkHarnessClient) notification.getValue();
```

Comment:
Again, not nice but unfortunately in this order we can't use proper type information.

Reply:
This is resolved now.